### PR TITLE
CIRC-1875 Filter request types based on item statuses

### PR DESCRIPTION
**Pull request by alexanderkurash**
Resolves [CIRC-1875](https://issues.folio.org/browse/CIRC-1875)

## Purpose
Available request types should be limited by both request policy and item status. Before the new endpoint (allowed-service-points) was introduced, FE was checking item statuses (ILR only). Now this logic needs to move to BE, foe both ILR and TLR.

## Approach
Add changes to rules execution code to preserve items that contributed to the inclusion of a particular policy into the result.

### CODE REVIEW

File: src/main/java/org/folio/circulation/infrastructure/storage/requests/RequestPolicyRepository.java [hunk 1]
```diff
@@ -72,23 +76,31 @@ public CompletableFuture<Result<RequestPolicy>> lookupRequestPolicy(Item item, U
...
-      .thenApply(r -> r.map(HashSet::new))
+      .collect(toMap(identity(), criteria -> List.of(criteria.getItem()), itemListsMergeOperator()));
+
+    return allOf(criteriaMap.keySet(),
+      criteria -> lookupRequestPolicyAndBuildContext(criteria, criteriaMap.get(criteria)))
+      .thenApply(r -> r.map(requestPolicyContexts -> requestPolicyContexts.stream()
+        .collect(toMap(RequestPolicyContext::getRequestPolicyId, RequestPolicyContext::getItems,
+          itemListsMergeOperator()))
+      ))
       .thenCompose(r -> r.after(this::lookupRequestPolicies));
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/infrastructure/storage/requests/RequestPolicyRepository.java:95]:
```suggestion
    return allOf(criteriaMap.entrySet(), entry -> lookupRequestPolicyId(entry.getKey())
        .thenApply(r -> r.map(id -> Pair.of(id, entry.getValue()))))
      .thenApply(r -> r.map(pair -> pair.stream()
        .collect(toMap(Pair::getKey, Pair::getValue, itemListsMergeOperator()))))
      .thenCompose(r -> r.after(this::lookupRequestPolicies));
```

File: src/main/java/org/folio/circulation/support/AsyncCoordinationUtil.java [hunk 2]
```diff
@@ -47,6 +48,29 @@ public static <T, R> CompletableFuture<List<Result<R>>> allResultsOf(
...
+    return allResultsOf(map, asyncAction)
+      .thenApply(Result::combineAll);
+  }
+
+  private static <T, R, S> CompletableFuture<List<Result<S>>> allResultsOf(
+    Map<T, R> map, BiFunction<T, R, CompletableFuture<Result<S>>> asyncAction) {
+
+    List<CompletableFuture<Result<S>>> futures =
+      map.keySet().stream()
+        .map(key -> asyncAction.apply(key, map.get(key)))
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/support/AsyncCoordinationUtil.java:67]: Maybe it's better to iterate over entry set? This way you we won't have to lookup value by key

File: src/test/java/api/requests/AllowedServicePointsAPITests.java [hunk 3]
```diff
@@ -81,96 +89,268 @@ void getFailsWithMultipleErrors() {
...
 
     var requesterId = usersFixture.steve().getId().toString();
-    var itemId = itemsFixture.basedUponNod().getId().toString();
-    var instanceId = itemsFixture.createMultipleItemsForTheSameInstance(2).get(0)
-      .getInstanceId().toString();
-    var cd1 = servicePointsFixture.cd1();
-    var cd2 = servicePointsFixture.cd2();
-    setRequestPolicyWithAllowedServicePoints(requestType, cd1.getId(), cd2.getId());
+    var items = itemsFixture.createMultipleItemForTheSameInstance(1,
+      List.of(ib -> ib.withStatus(itemStatus.getValue())));
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/AllowedServicePointsAPITests.java:125]: How about something simpler?
```suggestion
    var item = itemsFixture.basedUponNod(ib -> ib.withStatus(itemStatus.getValue()))
```

File: src/main/java/org/folio/circulation/support/AsyncCoordinationUtil.java (hunk 2 above)
Comment (alexanderkurash) [reply src/main/java/org/folio/circulation/support/AsyncCoordinationUtil.java]: done

File: src/main/java/org/folio/circulation/infrastructure/storage/requests/RequestPolicyRepository.java (hunk 1 above)
Comment (alexanderkurash) [reply src/main/java/org/folio/circulation/infrastructure/storage/requests/RequestPolicyRepository.java]: removed RequestPolicyContext

File: src/main/java/org/folio/circulation/infrastructure/storage/requests/RequestPolicyRepository.java [hunk 4]
```diff
@@ -72,23 +76,31 @@ public CompletableFuture<Result<RequestPolicy>> lookupRequestPolicy(Item item, U
...
   }
 
+  private BinaryOperator<List<Item>> itemListsMergeOperator() {
+    return (itemList1, itemList2) -> Stream.concat(itemList1.stream(), itemList2.stream())
+      .distinct()
+      .toList();
```
[the remark this answers is not in the document]
Comment (alexanderkurash) [reply src/main/java/org/folio/circulation/infrastructure/storage/requests/RequestPolicyRepository.java]: changed item lists to sets

**Review by OleksandrVidinieiev — APPROVED** [review]

**Review by roman-barannyk — APPROVED** [review]

**Pull request merged**